hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/procedures/ShowTimelineProcedure.scala

```diff
@@ -0,0 +1,381 @@
+    StructField("completed_time", DataTypes.StringType, nullable = true, Metadata.empty),
+    StructField("timeline_type", DataTypes.StringType, nullable = true, Metadata.empty),
+    StructField("rollback_info", DataTypes.StringType, nullable = true, Metadata.empty)
+  ))
+
+  def parameters: Array[ProcedureParameter] = PARAMETERS
+
+  def outputType: StructType = OUTPUT_TYPE
+
+  override def call(args: ProcedureArgs): Seq[Row] = {
+    super.checkArgs(PARAMETERS, args)
+
+    val tableName = getArgValueOrDefault(args, PARAMETERS(0))
+    val tablePath = getArgValueOrDefault(args, PARAMETERS(1))
+    val limit = getArgValueOrDefault(args, PARAMETERS(2)).get.asInstanceOf[Int]
+    val showArchived = getArgValueOrDefault(args, PARAMETERS(3)).get.asInstanceOf[Boolean]
+    val filter = getArgValueOrDefault(args, PARAMETERS(4)).get.asInstanceOf[String]
+    val startTime = getArgValueOrDefault(args, PARAMETERS(5)).get.asInstanceOf[String]
+    val endTime = getArgValueOrDefault(args, PARAMETERS(6)).get.asInstanceOf[String]
+
+    validateFilter(filter, outputType)
+
+    val basePath: String = getBasePath(tableName, tablePath)
+    val metaClient = createMetaClient(jsc, basePath)
+
+    val timelineEntries = getTimelineEntries(metaClient, limit, showArchived, startTime, endTime)
+
+    applyFilter(timelineEntries, filter, outputType)
+  }
+
+  override def build: Procedure = new ShowTimelineProcedure()
+
+  private def getTimelineEntries(metaClient: HoodieTableMetaClient,
+                                 limit: Int,
+                                 showArchived: Boolean,
+                                 startTime: String,
+                                 endTime: String): Seq[Row] = {
+
+    val instantInfoMap = buildInstantInfoFromTimeline(metaClient)
+
+    val activeRollbackInfoMap = getRolledBackInstantInfo(metaClient.getActiveTimeline, metaClient)
+    val archivedRollbackInfoMap = if (showArchived) {
+      getRolledBackInstantInfo(metaClient.getArchivedTimeline, metaClient)
+    } else {
+      Map.empty[String, List[String]]
+    }
+
+    val activeEntries = getTimelineEntriesFromTimeline(
+      metaClient.getActiveTimeline, "ACTIVE", metaClient, instantInfoMap, activeRollbackInfoMap, limit, startTime, endTime
+    )
+
+    val finalEntries = if (showArchived) {
+      val archivedEntries = getTimelineEntriesFromTimeline(
+        metaClient.getArchivedTimeline, "ARCHIVED", metaClient, instantInfoMap, archivedRollbackInfoMap, limit, startTime, endTime
+      )
+      val combinedEntries = (activeEntries ++ archivedEntries)
+        .sortWith((a, b) => a.getString(0) > b.getString(0))
+
+      if (startTime.trim.nonEmpty && endTime.trim.nonEmpty) {
+        combinedEntries
+      } else {
+        combinedEntries.take(limit)
+      }
+    } else {
+      if (startTime.trim.nonEmpty && endTime.trim.nonEmpty) {
+        activeEntries
+      } else {
+        activeEntries.take(limit)
+      }
+    }
```

Comment:
Maybe we need to redesign the use of limit. IMO, the limit should have nothing to do with the start time and end time.